migration of search syntax throws exception and generates empty/defunct table

### JabRef version

Latest development branch build (please note build date below)

### Operating system

GNU / Linux

### Details on version and operating system

JabRef 5.16--2024-09-10--ce9e34f Linux 6.10.8-200.fc40.x86_64 amd64  Java 21.0.2  JavaFX 22.0.2+4

### Checked with the latest development build (copy version output from About dialog)

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. open database with old search group syntax
2. migration assistant shows up :white_check_mark: 
3. agree to running the migration
4. then exception occurs
5. then migration seems to happening / running wheel shows up
6. then in different test of JabRef nothing happens, (wheel continues to run for many minutes) or a blank table is presented.


### Appendix

...
<details>

<summary>Log File</summary>

```
org.antlr.v4.runtime.misc.ParseCancellationException
	at org.jabref.merged.module@5.16.345/org.antlr.v4.runtime.BailErrorStrategy.recoverInline(Unknown Source)
	at org.jabref@5.16.345/org.jabref.search.SearchParser.name(Unknown Source)
	at org.jabref@5.16.345/org.jabref.search.SearchParser.comparison(Unknown Source)
	at org.jabref@5.16.345/org.jabref.search.SearchParser.expression(Unknown Source)
	at org.jabref@5.16.345/org.jabref.search.SearchParser.start(Unknown Source)
	at org.jabref@5.16.345/org.jabref.migrations.SearchToLuceneMigration.getStartContext(Unknown Source)
	at org.jabref@5.16.345/org.jabref.migrations.SearchToLuceneMigration.migrateToLuceneSyntax(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.importer.actions.SearchGroupsMigrationAction.migrateGroups(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.importer.actions.SearchGroupsMigrationAction.migrateGroups(Unknown Source)
	at java.base/java.util.Optional.ifPresent(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.importer.actions.SearchGroupsMigrationAction.performAction(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.importer.actions.OpenDatabaseAction.performPostOpenActions(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.LibraryTab.onDatabaseLoadingSucceed(Unknown Source)
	at org.jabref@5.16.345/org.jabref.gui.util.UiTaskExecutor.lambda$getJavaFXTask$3(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
	at javafx.base@22.0.2/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
	at javafx.base@22.0.2/javafx.event.Event.fireEvent(Unknown Source)
	at javafx.graphics@22.0.2/javafx.concurrent.EventHelper.fireEvent(Unknown Source)
	at javafx.graphics@22.0.2/javafx.concurrent.Task.fireEvent(Unknown Source)
	at javafx.graphics@22.0.2/javafx.concurrent.Task.setState(Unknown Source)
	at javafx.graphics@22.0.2/javafx.concurrent.Task$TaskCallable.lambda$call$1(Unknown Source)
	at javafx.graphics@22.0.2/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at javafx.graphics@22.0.2/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at javafx.graphics@22.0.2/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at javafx.graphics@22.0.2/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at javafx.graphics@22.0.2/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$10(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.antlr.v4.runtime.InputMismatchException
	... 33 more

```

</details>

--- a/src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java
+++ b/src/main/java/org/jabref/gui/groups/GroupDialogViewModel.java
@@ -26,6 +26,7 @@ import javafx.scene.paint.Color;
 import org.jabref.gui.DialogService;
 import org.jabref.gui.StateManager;
 import org.jabref.gui.icon.IconTheme;
+import org.jabref.gui.importer.actions.SearchGroupsMigrationAction;
 import org.jabref.gui.preferences.GuiPreferences;
 import org.jabref.gui.util.FileDialogConfiguration;
 import org.jabref.logic.auxparser.DefaultAuxParser;
@@ -318,6 +319,17 @@ public class GroupDialogViewModel {
                         groupHierarchySelectedProperty.getValue(),
                         searchGroupSearchTermProperty.getValue().trim(),
                         searchFlagsProperty.getValue());
+
+                if (currentDatabase.getMetaData().getGroupSearchSyntaxVersion().isEmpty()) {
+                    // If the syntax version for search groups is not present, it indicates that the groups
+                    // have not been migrated to the new syntax, or this is the first search group in the library.
+                    // If this is the first search group, set the syntax version to the new version.
+                    // Otherwise, it means that the user did not accept the migration to the new version.
+                    Optional<GroupTreeNode> groups = currentDatabase.getMetaData().getGroups();
+                    if (groups.filter(this::groupOrSubgroupIsSearchGroup).isEmpty()) {
+                        currentDatabase.getMetaData().setGroupSearchSyntaxVersion(SearchGroupsMigrationAction.VERSION_6_0_ALPHA);
+                    }
+                }
 
                 Optional<LuceneManager> luceneManager = stateManager.getLuceneManager(currentDatabase);
                 if (luceneManager.isPresent()) {
@@ -619,5 +631,17 @@ public class GroupDialogViewModel {
 
     public StringProperty texGroupFilePathProperty() {
         return texGroupFilePathProperty;
+    }
+
+    private boolean groupOrSubgroupIsSearchGroup(GroupTreeNode groupTreeNode) {
+        if (groupTreeNode.getGroup() instanceof SearchGroup) {
+            return true;
+        }
+        for (GroupTreeNode child : groupTreeNode.getChildren()) {
+            if (groupOrSubgroupIsSearchGroup(child)) {
+                return true;
+            }
+        }
+        return false;
     }
 }

--- a/src/main/java/org/jabref/gui/importer/actions/SearchGroupsMigrationAction.java
+++ b/src/main/java/org/jabref/gui/importer/actions/SearchGroupsMigrationAction.java
@@ -13,6 +13,8 @@ import org.jabref.model.groups.GroupTreeNode;
 import org.jabref.model.groups.SearchGroup;
 import org.jabref.model.search.SearchFlags;
 
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+
 /**
  * This action checks whether the syntax for SearchGroups is the new one.
  * If not we ask the user whether to migrate.
@@ -54,18 +56,26 @@ public class SearchGroupsMigrationAction implements GUIPostOpenAction {
             return;
         }
 
-        parserResult.getMetaData().getGroups().ifPresent(this::migrateGroups);
+        parserResult.getMetaData().getGroups().ifPresent(groupTreeNode -> migrateGroups(groupTreeNode, dialogService));
         parserResult.getMetaData().setGroupSearchSyntaxVersion(VERSION_6_0_ALPHA);
         parserResult.setChangedOnMigration(true);
     }
 
-    private void migrateGroups(GroupTreeNode node) {
+    private void migrateGroups(GroupTreeNode node, DialogService dialogService) {
         if (node.getGroup() instanceof SearchGroup searchGroup) {
-            String luceneSearchExpression = SearchToLuceneMigration.migrateToLuceneSyntax(searchGroup.getSearchExpression(), searchGroup.getSearchFlags().contains(SearchFlags.REGULAR_EXPRESSION));
-            searchGroup.setSearchExpression(luceneSearchExpression);
+            try {
+                String luceneSearchExpression = SearchToLuceneMigration.migrateToLuceneSyntax(searchGroup.getSearchExpression(), searchGroup.getSearchFlags().contains(SearchFlags.REGULAR_EXPRESSION));
+                searchGroup.setSearchExpression(luceneSearchExpression);
+            } catch (ParseCancellationException e) {
+                Optional<String> luceneSearchExpression = dialogService.showInputDialogWithDefaultAndWait(
+                        Localization.lang("Search group migration failed"),
+                        Localization.lang("The search group '%0' could not be migrated. Please enter the new search expression.", searchGroup.getName()),
+                        searchGroup.getSearchExpression());
+                luceneSearchExpression.ifPresent(searchGroup::setSearchExpression);
+            }
         }
         for (GroupTreeNode child : node.getChildren()) {
-            migrateGroups(child);
+            migrateGroups(child, dialogService);
         }
     }
 }

--- a/src/main/resources/l10n/JabRef_en.properties
+++ b/src/main/resources/l10n/JabRef_en.properties
@@ -270,9 +270,11 @@ Dynamically\ group\ entries\ by\ searching\ a\ field\ for\ a\ keyword=Dynamicall
 Each\ line\ must\ be\ of\ the\ following\ form\:\ \'tab\:field1;field2;...;fieldN\'.=Each line must be of the following form\: 'tab\:field1;field2;...;fieldN'.
 
 Search\ groups\ migration\ of\ %0=Search groups migration of %0
-The\ search\ groups\ syntax\ is\ outdated.\ Do\ you\ want\ to\ migrate\ to\ the\ new\ syntax?= The search groups syntax is outdated. Do you want to migrate to the new syntax?
+The\ search\ groups\ syntax\ is\ outdated.\ Do\ you\ want\ to\ migrate\ to\ the\ new\ syntax?=The search groups syntax is outdated. Do you want to migrate to the new syntax?
 Migrate=Migrate
 Keep\ as\ is=Keep as is
+Search\ group\ migration\ failed=Search group migration failed
+The\ search\ group\ '%0'\ could\ not\ be\ migrated.\ Please\ enter\ the\ new\ search\ expression.=The search group '%0' could not be migrated. Please enter the new search expression.
 Edit=Edit
 
 Edit\ file\ type=Edit file type

